### PR TITLE
Implement <Table /> component

### LINKED ISSUE #60: Implement <Table /> component
Implement a new `<Table />` component or adjust the existing one. 
Please refer to Figma for the details.

Relates to: 
- [x] #45 
- [ ] #47 
- [x] #52 
- [ ] #63 
- [ ] #78 
- [x] #77 

The table should have sub-menus:

Image

Image

("No controls" table)
Image

Also, there should be the ability to make the sub-menu block fixed: 

("No controls" table)
Image

Also, there should be the ability to mark a row as disabled: 

Image

### LINKED ISSUE #78: Implement table sorting
Image

We need to add the ability to sort by the table's column values. Only one column at a time can be sorted. 
The sorting should be ascending and descending. 
The sorting doesn't modify the data displayed in the table but calls a function to be handled by a parent component.

Relates to #60

### PR DESCRIPTION
**Pull request by react-developer-ssa-group**
[Implement Table component](https://github.com/ssagroup/ui-kit/issues/60)
[Implement table sorting](https://github.com/ssagroup/ui-kit/issues/78)

Closes #60 
Closes #78

### CODE REVIEW

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 4  | ❗       |
    


*❗ storybook/components-icons--icons__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/components-icons--icons__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/components-icons--icons__[w390px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/components-icons--icons__[w900px]*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 8       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 8  | ❗       |
    


<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>❗ custom/widgets-meal-nutrients--hint-opened__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-meal-nutrients--hint-opened__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-meal-nutrients--hint-opened__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ custom/widgets-meal-nutrients--hint-opened__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-calories--default__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-heart-rate--default__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-heart-rate--default__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/widgets-steps-counter--default__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 6  | ❗       |
    


<details>
<summary>❗ storybook/components-icons--icons__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-icons--icons__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-icons--icons__[w390px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/components-icons--icons__[w900px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/utility-responsive-image--responsive-image__[w1440px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>



<details>
<summary>❗ storybook/utility-responsive-image--responsive-image__[w1920px] (click to expand)</summary>

| Before | After |
| --- | --- |
|  |  |
</details>

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 10       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 5  | ❗       |
    


<details>
<summary>🟡 custom/widgets-table-bots--no-control-orders--menu-opened_[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 custom/widgets-table-bots--no-control-orders--run-reason-modal_[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--no-control-orders__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-table-bots--table-bots__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



*❗ custom/widgets-meal-nutrients--hint-opened__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-meal-nutrients--hint-opened__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-meal-nutrients--hint-opened__[w390px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/widgets-calories--default__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/widgets-heart-rate--default__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |

File: packages/widgets/src/components/TableBots/TableBots.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+      expect(consoleMock).toHaveBeenCalledWith(
+        'Submitting... ' +
+          JSON.stringify({
+            reason: 'reason',
+          }),
+      );
+    });
+  });
+
+  it('Should be clicked by the more dropdown item', async () => {
```
Comment (arthur-kashaiev-ssa-group) [inline packages/widgets/src/components/TableBots/TableBots.spec.tsx:77]: The name of this test doesn't correspond to its content

Comment (arthur-kashaiev-ssa-group) [reply packages/widgets/src/components/TableBots/TableBots.spec.tsx:77]: The same is true for the test case on the line `#109`.

File: packages/widgets/src/components/TableBots/stories/TableBots/components/Actions/ActionRun.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import { actionsIcons } from './consts';
+
+export const ActionRun = ({ row }: { row: TableBotItem }) => {
+  const { status, isDisabled } = row;
+  const ActionIcon = actionsIcons[status];
+
+  const [open, setOpen] = useState(false);
+  const { register, handleSubmit, resetField } = useForm<FieldValues>();
+
+  // const onSubmit = handleSubmit((data) => {
```
Comment (arthur-kashaiev-ssa-group) [inline packages/widgets/src/components/TableBots/stories/TableBots/components/Actions/ActionRun.tsx:23]: Remove the comment

File: packages/widgets/src/components/TableBots/types.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  currency: string;
+  isIncreasing: boolean;
+}
+
+export interface ROI {
+  amount: number;
+  isIncreasing: boolean;
+}
+
+export interface TableBotItem {
```
Comment (arthur-kashaiev-ssa-group) [inline packages/widgets/src/components/TableBots/types.ts:16]: Should this interface and `NoControlOrdersItem` be here or deeper on the stories level?

**Review by arthur-kashaiev-ssa-group — CHANGES_REQUESTED** [review]

File: packages/core/src/components/Icon/icons/MoreVertical.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { SVGProps } from '@components/Icon/types';
+
+export const MoreVertical = ({ fill = '#000', size = 24 }: SVGProps) => (
```
[the remark this answers is not in the document]
Comment (react-developer-ssa-group) [reply packages/core/src/components/Icon/icons/MoreVertical.tsx:3]: We have already `More` icon. The name "MoreVertical" goes from "More" icon name

File: packages/widgets/src/components/TableBots/types.ts (hunk 3 above)
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/TableBots/types.ts:16]: Agree, moved

File: packages/widgets/src/components/TableBots/TableBots.spec.tsx (hunk 1 above)
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/TableBots/TableBots.spec.tsx]: Changed, thanks

File: packages/core/src/components/Tag/Tag.tsx [hunk 5]
```diff
@@ -33,8 +33,13 @@ const mapColors: MainColors = {
   yellowWarm: [yellowWarm, yellowWarmBorder],
 };
 
-const Tag = ({ color = 'purple', size = 'medium', children }: TagsProps) => (
-  <Badge color={color} size={size} css={mapColors[color]}>
+const Tag = ({
+  color = 'purple',
+  size = 'medium',
+  extraCSS,
```
[the remark this answers is not in the document]
Comment (react-developer-ssa-group) [reply packages/core/src/components/Tag/Tag.tsx]: Done

File: packages/widgets/src/components/TableBots/TableBot.tsx [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import styled from '@emotion/styled';
+import { Table } from '@ssa-ui-kit/core';
+
+export const TableBot = styled(Table)`
```
[the remark this answers is not in the document]
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/TableBots/TableBot.tsx]: Done

File: packages/widgets/src/components/TableBots/TableBots.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    getByText('Bot name 1');
+    getAllByText('27.07.23');
+    getByText('Bittrex');
+    getByText('Pending');
+    getAllByText('ETH/USD');
+    getByText('-70 USD');
+    getByText('-10%');
+  });
+
+  it('Should be alert showed after clicking a table row', () => {
```
[the remark this answers is not in the document]
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/TableBots/TableBots.spec.tsx]: Done

File: packages/widgets/src/components/TableBots/TableBots.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      'tbody > tr:nth-child(6) > td:last-child button',
+    );
+
+    fireEvent.click(button as Node);
+
+    getByText('Run Reason');
+  });
+
+  it('Should be submitted "Run Reason" modal', async () => {
+    const consoleMock = jest.spyOn(console, 'log');
```
[the remark this answers is not in the document]
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/TableBots/TableBots.spec.tsx]: Done

File: packages/widgets/src/components/TableBots/stories/TableBots/components/Actions/ActionRun.tsx (hunk 2 above)
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/TableBots/stories/TableBots/components/Actions/ActionRun.tsx]: Done

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 10       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 4  | ❗       |
    


<details>
<summary>🟡 custom/widgets-bots-table--no-control-orders--menu-opened_[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 custom/widgets-bots-table--no-control-orders--run-reason-modal_[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--bots-table__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--bots-table__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--bots-table__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--bots-table__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--no-control-orders__[w1440px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--no-control-orders__[w1920px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--no-control-orders__[w390px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



<details>
<summary>🟡 storybook/widgets-bots-table--no-control-orders__[w900px] (click to expand)</summary>

| Image |
| --- |
|  |
</details>



*❗ custom/widgets-user-profile--is-opened__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/widgets-calories--default__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/widgets-heart-rate--default__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/widgets-heart-rate--default__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |

File: packages/widgets/src/components/BotsTable/types.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  pnl: PNL;
+  roi: ROI;
+  isDisabled: boolean;
+}
+
+export interface BotsTableProps {
+  children: React.ReactElement<React.PropsWithChildren<BotsTableItem>>[];
+  columns: string[];
+  className?: string;
+  css?: Interpolation<Theme>;
```
Comment (arthur-kashaiev-ssa-group) [inline packages/widgets/src/components/BotsTable/types.ts:32]: Do you need both `className` and `css` here? 
`emotion` converts `css` into `className`.

File: packages/core/src/components/Table/Table.spec.tsx [hunk 10]
```diff
@@ -38,4 +41,45 @@ describe('Table', () => {
     expect(table).toBeInTheDocument();
     expect(cells.length).toEqual(9);
   });
+
+  it('Render Sorting table', () => {
```
Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Table/Table.spec.tsx:45]: Should we move test cases related to the sorting into a nested `describe` block? e.g. 
```
describe('Table', () => {
     describe('SortingTable', () => {
       // ...
     })
  })
```

File: packages/core/src/components/Table/Table.spec.tsx [hunk 11]
```diff
@@ -38,4 +41,45 @@ describe('Table', () => {
...
+    render(<TableSortingComponent />);
+
+    const cells = screen.queryAllByRole('cell');
+
+    expect(screen.getByRole('table')).toBeInTheDocument();
+    expect(screen.getByTitle('Arrow Down')).toBeInTheDocument();
+    expect(cells.length).toEqual(33);
+  });
+
+  it('Should sort changing worked', () => {
```
Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Table/Table.spec.tsx:55]: `it should sort changing worked` -> `it should change sorting correctly`.

File: packages/core/src/components/Table/components/TableSortingComponent.tsx [hunk 12]
```diff
@@ -0,0 +1,82 @@
+import { useState } from 'react';
```
Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Table/components/TableSortingComponent.tsx:1]: We don't use the nested `/components` directory in the project. Consider creating a sub-directory name by the component, e.g. `TableSorting`, or just put the component into the `Table/` directory.

File: packages/core/src/components/Table/components/TableSortingComponent.tsx [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import {
+  SortInfo,
+  defaultSort,
+  RowKeys,
+  headers,
+  dataForTableSorting,
+} from './mock';
+import { capitalizeFirstLetter } from './utils';
+
+export const TableSortingComponent = ({
```
Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Table/components/TableSortingComponent.tsx:18]: Don't use `Component` postfix in component names.

Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Table/components/TableSortingComponent.tsx:18]: Let's call the component `SortableTable`

File: packages/core/src/components/Table/components/TableSortingComponent.tsx (hunk 12 above)
Comment (arthur-kashaiev-ssa-group) [reply packages/core/src/components/Table/components/TableSortingComponent.tsx:1]: If it's for Stories only, them move into the `./stories` directory as we usually do.

File: packages/core/src/components/Table/components/mock.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
+export const dataForTableSorting = Array.from({ length: 10 }).map(
+  (_, index) => ({
+    title: `title ${index + 1}`,
+    value: `value ${index + 1}`,
+    description: `description ${index + 1}`,
+  }),
+);
+
+export type DataRow = (typeof dataForTableSorting)[0];
```
Comment (arthur-kashaiev-ssa-group) [inline packages/core/src/components/Table/components/mock.ts:9]: I'm not quite sure if this file is for Stories only or not. If not, let's use the `./types` file as usual.

File: packages/widgets/src/components/BotsTable/BotsTable.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+      expect(consoleMock).toHaveBeenCalledWith(
+        'Submitting... ' +
+          JSON.stringify({
+            reason: 'reason',
+          }),
+      );
+    });
+  });
+
+  it('Should be clicked by the three dots button', async () => {
```
Comment (arthur-kashaiev-ssa-group) [inline packages/widgets/src/components/BotsTable/BotsTable.spec.tsx:77]: `it should be clicked by the three dots button` means that the table was clicked by the button. I mean, not by a user but by the button, literally.
Use something like this: 
> it should have the menu item clicked

Comment (arthur-kashaiev-ssa-group) [reply packages/widgets/src/components/BotsTable/BotsTable.spec.tsx:77]: Here and in the similar test case below

**Review by arthur-kashaiev-ssa-group — CHANGES_REQUESTED** [review]

File: packages/widgets/src/components/BotsTable/types.ts (hunk 9 above)
Comment (react-developer-ssa-group) [reply packages/widgets/src/components/BotsTable/types.ts:32]: No, we need only `className` prop

File: packages/core/src/components/Table/Table.spec.tsx (hunk 10 above)
Comment (react-developer-ssa-group) [reply packages/core/src/components/Table/Table.spec.tsx:45]: Yes, we can move

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 2  | ❗       |
    


*❗ storybook/components-indicator--with-content__[w900px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ storybook/utility-responsive-image--responsive-image__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |

**Comment by lost-pixel-signals** [comment]
| Shot Type       | Count                | Status   |
| -------------   | -------------------- | -------- |
| Added           | 0       | 🟡       |
| Deleted         | 0     | ❌       |
| Regression      | 5  | ❗       |
    


*❗ custom/widgets-meal-nutrients--hint-opened__[w390px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-navbar-opened__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-progressInfo__[w1440px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-progressInfo__[w1920px]*: 


| Before | After |
| --- | --- |
|  |  |



*❗ custom/widgets-progressInfo__[w900px]*: 


| Before | After |
| --- | --- |
|  |  |

**Review by arthur-kashaiev-ssa-group — APPROVED** [review]

**Pull request merged**